app/tests/spec/lib/crypto/recovery-keys.js

```diff
@@ -0,0 +1,145 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  const {assert} = require('chai');
+  const RecoveryKeys = require('lib/crypto/recovery-keys');
+  const validate = require('../../../../scripts/lib/validate');
+
+  const uid = 'aaaaabbbbbcccccdddddeeeeefffff00';
+  const recoveryKey = 'AFAUEQ2EIVDEOSCJJJFUYTKOJ5IFCUST';
+  const recoveryData = {
+    kB: '000000111111222222333333444444555555666666777777888888999999ABCD'
+  };
+  const expectedRecoveryKeyId = 'edc243a821582ee9e979583be9989ee7';
+  const expectedRecoveryEnc = '6737e3c3c678f0dc67abfb853115e7c702933795a3662a99d1b27b6e645dc506';
+  // const expectedBundle = 'eyJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiZGlyIiwia2lkIjoiODE4NDIwZjBkYTU4ZDIwZjZhZTR' +
+  //   'kMmM5YmVhYjkyNTEifQ..D29EXHp8ubLvftaZ.xHJd2Nl2Uco2RyywYPLkUU7fHpgO2FztY12Zjpq1ffiyLRIUcQVfmiNC6aMiHB' +
+  //   'l7Hp-lXEbb5mR1uXHrTH9iRXEBVaAfyf9KEAWOukWGVSH8EaOkr7cfu2Yr0K93Ec8glsssjiKp8NGB8VKTUJ-lmBv2cIrG68V4eTUVDo' +
+  //   'DhMbXhrF-Mv4JNeh338pPeatTnyg.Ow2bhEYWxzxfSPMxVwKmSA';
+  // const encryptionOptions = {
+  //   iv: 'eeddccbbaa99887766554433'
```

Comment:
I was able to get a closer match to the test vectors here, by passing in the following encryption options:

```{ contentAlg: 'A256GCM', format: 'compact', iv: jose.util.base64url.encode('eeddccbbaa99887766554433', 'hex') ```

With the IV passed in this this way, the *iv* and *ciphertext* portions of the JWE match between javascript and my python test code.  I get the following from javascript:

```
eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwia2lkIjoiS2tKLUkxa2JGQmU3cVN5RkxHRG5xZmxjZmpUX3RzTDJRTF8tTnJkLXhvayJ9..7t3Mu6qZiHdmVUQz._W-wCvuGOwqC7wMBZ5n3v48FDS0HyXvGVcmPpqNrufftkclyacWbaeHiaEfDmNTm8PqFGu69pt1IELX6zo7cdQaruGgo9CFnq0E.93WffJuqaCqZ4CCIkigMuA
```

Versus the following from python:

```
 eyJhbGciOiAiZGlyIiwgImVuYyI6ICJBMjU2R0NNIn0..7t3Mu6qZiHdmVUQz._W-wCvuGOwqC7wMBZ5n3v48FDS0HyXvGVcmPpqNrufftkclyacWbaeHiaEfDmNTm8PqFGu69pt1IELX6zo7cdQaruGgo9CFnq0E.V2jqPbQZcpAEseEJOfRLYg
```

Comparing the dotted components of each, I see `<different>..<same>.<same>.<different>`.  The final component is the authentication tag, which we should expect to be different if there's a difference in the first component.  So the only thing that really differs is the first component.

Decoding the first component from the javascript version:

```
> jose.util.base64url.decode('eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwia2lkIjoiS2tKLUkxa2JGQmU3cVN5RkxHRG5xZmxjZmpUX3RzTDJRTF8tTnJkLXhvayJ9').toString()
'{"alg":"dir","enc":"A256GCM","kid":"KkJ-I1kbFBe7qSyFLGDnqflcfjT_tsL2QL_-Nrd-xok"}'
```

Decoding the first component from the pythonversion:

```
> jose.util.base64url.decode('eyJhbGciOiAiZGlyIiwgImVuYyI6ICJBMjU2R0NNIn0').toString()
'{"alg": "dir", "enc": "A256GCM"}'
```

So the difference is simply that node-jose is formatting the protected header differently to how the python version does it.

I'm not sure what it would take for us to get them producing bit-identical output, but a good interim state might be to use the fixed IV as above, split the resulting JWE into its dotted components, and assert that the ones we know should match actually match.  Does that make sense?

Reply:
> jose.util.base64url.encode('eeddccbbaa99887766554433', 'hex')

Ah, in my original I didn't base64 encode.

> a good interim state might be to use the fixed IV as above, split the resulting JWE into its dotted components, and assert that the ones we know should match actually match

This makes sense to me. I think it is a decent interim until we decided how to make them identical.

Reply:
By default `node-jose` includes the `kid` in the resulting JWE/JWS headers, which is generated using "JWK thumbprint" if one is not present in the source JWK JSON.  See comment on line 117 of `app/scripts/lib/crypto/recovery-keys.js` for how to omit the `kid`.